File: rec.js
function caminhoDasPedras(caminho, cansado = false, pos = 0) {
	if (caminho[pos] == "0" || pos >= caminho.length) {
		return 0;
	}

	if (caminho[pos] == "m" && pos == caminho.length - 1) {
		return 1;
	}

	let sum = 0;

	for (let i = 1; i <= 3 - cansado; i++) {
		sum += caminhoDasPedras(caminho, i == 3, pos + i);
	}

	return sum;
}

const args = process.argv.slice(2);

if (args.length != 1) {
	console.log("Uso: node rec.js <caminho>");
	process.exit(1);
}

console.log(caminhoDasPedras(args[0]));

File: src/bin/rec.rs
fn caminho_das_pedras(caminho: &str, cansado: bool) -> usize {
    match caminho.chars().next() {
        Some('0') | None => 0,
        Some('m') if caminho.len() == 1 => 1,
        _ => (1..=caminho.len().min(3 - cansado as usize))
            .map(|x| caminho_das_pedras(&caminho[x..], x == 3))
            .sum(),
    }
}

fn main() {
    let args = std::env::args().collect::<Vec<_>>();

    if args.len() != 2 {
        println!("Uso: {} <caminho>", args[0]);
        return;
    }

    println!("{}", caminho_das_pedras(&args[1], false));
}
